tests/matrix_free/matrix_vector_rt_common.h

```diff
@@ -0,0 +1,238 @@
+    mf_data.reinit(dof, constraints, quad, data);
+  }
+
+  // create vector with random entries
+  Vector<Number> solution, initial_condition;
+
+  mf_data.initialize_dof_vector(solution);
+  mf_data.initialize_dof_vector(initial_condition);
+
+  for (unsigned int i = 0; i < dof.n_dofs(); ++i)
+    {
+      if (constraints.is_constrained(i))
+        continue;
+      initial_condition[i] = random_value<Number>();
+    }
+
+  // MatrixFree solution
+  MatrixFreeTest<dim, fe_degree, fe_degree + 2, Number> mf(mf_data);
+  mf.test_functions(solution, initial_condition);
+
+
+  SparsityPattern        sp;
+  SparseMatrix<double>   system_matrix;
+  DynamicSparsityPattern dsp(dof.n_dofs(), dof.n_dofs());
+  DoFTools::make_sparsity_pattern(dof, dsp);
+  sp.copy_from(dsp);
+  system_matrix.reinit(sp);
+
+  FEValues<dim> fe_val(dof.get_fe(),
+                       Quadrature<dim>(mf_data.get_quadrature(0)),
+                       update_values | update_gradients | update_JxW_values);
+
+
+  const unsigned int dofs_per_cell = fe_val.get_fe().dofs_per_cell;
+  std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+  FullMatrix<double> local_matrix(dofs_per_cell, dofs_per_cell);
+
+
+  const FEValuesExtractors::Vector velocities(0);
+
+  // Assemble matrix
+  for (const auto &cell : dof.active_cell_iterators())
+    {
+      fe_val.reinit(cell);
+      local_matrix = 0;
+
+      for (const auto q : fe_val.quadrature_point_indices())
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          {
+            const Tensor<1, dim> phi_i = fe_val[velocities].value(i, q) * 10.;
+            const Tensor<2, dim> grad_phi_i = fe_val[velocities].gradient(i, q);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                const Tensor<1, dim> phi_j = fe_val[velocities].value(j, q);
+                const Tensor<2, dim> grad_phi_j =
+                  fe_val[velocities].gradient(j, q);
+
+                local_matrix(i, j) +=
+                  (phi_j * phi_i + scalar_product(grad_phi_i, grad_phi_j)) *
+                  fe_val.JxW(q);
+              }
+          }
+      cell->get_dof_indices(local_dof_indices);
+      for (unsigned int i = 0; i < dofs_per_cell; ++i)
+        for (unsigned int j = 0; j < dofs_per_cell; ++j)
+          system_matrix.add(local_dof_indices[i],
+                            local_dof_indices[j],
+                            local_matrix(i, j));
+    }
+
+  Vector<Number> ref(solution.size());
+
+  // Compute reference
+  system_matrix.vmult(ref, initial_condition);
+
+  ref -= solution;
+
+  const double diff_norm = ref.linfty_norm() / solution.linfty_norm();
+  deallog << "Norm of difference: " << diff_norm << std::endl << std::endl;
+}
+
+
+int
+main()
+{
+  initlog();
+
+  {
+    deallog.push("2d");
+    test<2, 2>();
+    test<2, 3>();
+    deallog.pop();
+    deallog.push("3d");
+    test<3, 2>();
+    test<3, 3>();
```

Comment:
It's just not at all as slow as computing the face integrals..

Reply:
It would be interesting to see why the face integrals (reference solution) are so horribly slow. Either way, the present contributions will bring the RT evaluation to a much more usable state especially in 3D and with degree larger than 1, even though we are still missing the last factor of 2 (even-odd decomposition, maybe some small notes).